tests/Fixer/ClassNotation/ClassAttributesSeparationFixerTest.php

```diff
@@ -0,0 +1,1094 @@
+    }
+
+    public function provideMessyWhitespacesCases()
+    {
+        return [
+            [
+                "<?php\r\nclass SomeClass\r\n{\r\n    // comment\n\n    public function echoA()\r\n    {\r\n        echo 'a';\r\n    }\r\n}\r\n",
+                "<?php\r\nclass SomeClass\r\n{\r\n    // comment\n\n\n    public function echoA()\r\n    {\r\n        echo 'a';\r\n    }\r\n}\r\n",
+            ],
+            [
+                "<?php\r\nclass SomeClass\r\n{\r\n    // comment\r\n\r\n    public function echoA()\r\n    {\r\n        echo 'a';\r\n    }\r\n}\r\n",
+                "<?php\r\nclass SomeClass\r\n{\r\n    // comment\r\n\r\n\r\n    public function echoA()\r\n    {\r\n        echo 'a';\r\n    }\r\n}\r\n",
+            ],
+        ];
+    }
+
+    /**
+     * @param string      $expected
+     * @param null|string $input
+     * @param array       $config
+     *
+     * @dataProvider provideConfigCases
+     */
+    public function testWithConfig($expected, $input = null, array $config = [])
+    {
+        $this->fixer->configure($config);
+        $this->doTest($expected, $input);
+    }
+
+    public function provideConfigCases()
+    {
+        return [
+            [
+                '<?php
+                    class A
+                    {
+                        private $a = null;
+
+                        public $b = 1;
+
+                    function A(){}}
+                ',
+                '<?php
+                    class A
+                    {
+                        private $a = null;
+                        public $b = 1;
+
+
+
+                    function A(){}}
+                ',
+                ['elements' => ['property']],
+            ],
+            [
+                '<?php
+                    class A
+                    {
+                        const A = 1;
+
+                        const THREE = ONE + self::TWO; /* test */ # test
+
+                        const B = 2;
+}
+                ',
+                '<?php
+                    class A
+                    {
+
+                        const A = 1;
+                        const THREE = ONE + self::TWO; /* test */ # test
+                        const B = 2;}
+                ',
+                ['elements' => ['const']],
+            ],
+        ];
+    }
+
+    /**
+     * @param string      $expected
+     * @param null|string $input
+     *
+     * @dataProvider provideFix70BigCases
+     * @requires PHP 7.0
+     * @coversNothing test cases here are too big to be executed with code coverage under phpdbg
+     */
+    public function testFix70Big($expected, $input = null)
+    {
+        $this->doTest($expected, $input);
+    }
+
+    public function provideFix70BigCases()
+    {
+        $to = $from = '<?php ';
+
+        for ($i = 0; $i < 15; ++$i) {
+            $from .= sprintf('class A%d{public function AA%d(){return new class {public function BB%d(){}};}public function otherFunction%d(){}}', $i, $i, $i, $i);
```

Comment:
note to self: %1$d should be used and not multiple $i as args.
but more important, this blows up the memory usage of the fixer to extreme values

Reply:
actually, string interpolation is better than sprintf on php 7

Reply:
3199 is supposed to fix it? if so, please cherry-pick it to this PR and revert my "big test" so we will see would it work indeed ;)